ecosystem/typescript/sdk_v2/src/bcs/serializer.ts

```diff
@@ -0,0 +1,246 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+/* eslint-disable no-bitwise */
+import {
+  MAX_U128_BIG_INT,
+  MAX_U16_NUMBER,
+  MAX_U32_NUMBER,
+  MAX_U64_BIG_INT,
+  MAX_U8_NUMBER,
+  MAX_U256_BIG_INT,
+} from "./consts";
+import { AnyNumber, Uint16, Uint32, Uint8 } from "./types";
+
+export interface Serializable {
+  serialize(serializer: Serializer): void;
+}
+
+export class Serializer {
+  private buffer: ArrayBuffer;
+
+  private offset: number;
+
+  constructor() {
+    this.buffer = new ArrayBuffer(64);
+    this.offset = 0;
+  }
+
+  private ensureBufferWillHandleSize(bytes: number) {
+    while (this.buffer.byteLength < this.offset + bytes) {
+      const newBuffer = new ArrayBuffer(this.buffer.byteLength * 2);
+      new Uint8Array(newBuffer).set(new Uint8Array(this.buffer));
+      this.buffer = newBuffer;
+    }
+  }
+
+  protected appendToBuffer(values: Uint8Array) {
+    this.ensureBufferWillHandleSize(values.length);
+    new Uint8Array(this.buffer, this.offset).set(values);
+    this.offset += values.length;
+  }
+
+  private serializeWithFunction(
+    fn: (byteOffset: number, value: number, littleEndian?: boolean) => void,
```

Comment:
I think BCS is only little endian?

Reply:
so the reason it has this is so the function signature matches the `Dataview.setUintX` functions later:

like you noticed:

```typescript
    fn: (byteOffset: number, value: number, littleEndian?: boolean) => void,
    // ...
    fn.apply(dv, [0, value, true]);
```
but if you see this:

```typescript
  serializeU16(value: Uint16) {
    this.serializeWithFunction(DataView.prototype.setUint16, 2, value);
  }
```

the `DataView.prototype.setUint16` is going to have a function signature with the `littleEndian` flag as the 3rd arg